### PR TITLE
US000 - Cardstack 2.0 - Enhancements

### PR DESCRIPTION
**Pull request by mohasin-nadaf**

### CODE REVIEW

File: src/common/cardstack/types.ts [hunk 1]
```diff
@@ -215,3 +240,18 @@ export interface AlCardstackPage<EntityType=any>
...
+{
+    property:AlCardstackPropertyDescriptor;                 //  The property being filtered on
+    propField:string;                                       //  The DTO field the property's value can be found on
+    values:AlCardstackValueDescriptor[];                    //  The values being filtered for (descriptors)
+    rawValues:any[];                                        //  The values being filtered for (raw values)
+
+    /**
+     *  The default callback will return true ("match") if any of the values in the filter match a given entity's value for that field, but it can be overridden to provide custom logic.
+     */
+    callback:{(entity:EntityType,properties:PropertyType,filter:AlCardstackActiveFilter<EntityType,PropertyType>):boolean};
```
Comment (mcnielsen) [inline src/common/cardstack/types.ts:256]: It would be good to test a custom callback on a filter, too, come to think of it.

Comment (mohasin-nadaf) [reply src/common/cardstack/types.ts:256]: We found today the need of customCallback for filter in portero app as some of the values are of type array.

File: src/common/cardstack/al-cardstack-view.ts [hunk 2]
```diff
@@ -273,42 +291,92 @@ export abstract class AlCardstackView< EntityType=any,
...
+            console.log("Could not find existing filter" );
+            return;
         }
-        delete this.activeFilters[propertyName][descriptor.valueKey];
-        if ( Object.keys( this.activeFilters[propertyName] ).length === 0 ) {
-            delete this.activeFilters[propertyName];
-        }
-        this.continuation = undefined;
 
+        console.log("Removing activeFilter from ", vDescriptor );
```
[the remark this answers is not in the document]
Comment (mohasin-nadaf) [reply src/common/cardstack/al-cardstack-view.ts]: Done, in state of updating its missed

File: src/common/cardstack/al-cardstack-view.ts [hunk 3]
```diff
@@ -78,10 +84,11 @@ export abstract class AlCardstackView< EntityType=any,
                 const characteristics = await this.generateCharacteristics();
                 this.normalizeCharacteristics( characteristics );
                 if (this.localPagination) {
+                    console.warn("WARNING: please do not use the localPagination property of AlCardstackView; it is deprecated.  Instead, use the localPagination property of the characteristics provided to the cardstack view." );
                     this.fillPropertiesReduceFilters();
```
[the remark this answers is not in the document]
Comment (mohasin-nadaf) [reply src/common/cardstack/al-cardstack-view.ts:87]: I modified this block with `characteristics.localPagination` and this property is kind of just blank initialization for reduceAllFilter method which might be optional.
We found its use case for al-notifications-list though

**Comment by mohasin-nadaf** [comment]
test it

**Comment by mohasin-nadaf** [comment]
manualpublish

**Comment by mohasin-nadaf** [comment]
manualpublish

**Comment by mohasin-nadaf** [comment]
test it

**Pull request merged**